File: apps/youtube/extensions/speed/src/main/java/app/reseam/youtube/speed/PlaybackSpeedDialogButton.java
// SPDX-FileCopyrightText: 2026 AunAli K. <hello@auna.li>
// SPDX-FileCopyrightText: 2026 ReVanced contributors
// SPDX-License-Identifier: GPL-3.0-or-later

package app.reseam.youtube.speed;

import android.view.View;

import java.text.DecimalFormat;

import app.reseam.youtube.controls.PlayerControlButton;
import app.reseam.youtube.core.Logger;
import app.reseam.youtube.core.Settings;
import app.reseam.youtube.video.VideoInformation;

/** A bottom-player button showing the current speed; it opens the speed sheet and resets on long press. */
public final class PlaybackSpeedDialogButton {
    private static final DecimalFormat speedFormatter = new DecimalFormat();
    private static PlayerControlButton button;

    static {
        speedFormatter.setMinimumFractionDigits(1);
        speedFormatter.setMaximumFractionDigits(2);
    }

    private PlaybackSpeedDialogButton() {}

    public static void initialize(View controls) {
        try {
            button = new PlayerControlButton(controls,
                    "reseam_playback_speed_button_container",
                    "reseam_playback_speed_button",
                    "reseam_playback_speed_button_text",
                    () -> Settings.getBoolean("playback_speed_dialog_button", false),
                    view -> {
                        if (Settings.getBoolean("restore_old_playback_speed_menu", false)) {
                            CustomPlaybackSpeed.showOldPlaybackSpeedMenu();
                        } else {
                            CustomPlaybackSpeed.showModernDialog(view.getContext());
                        }
                    },
                    view -> {
                        // Resets to the remembered default, or to 1.0x when already there, unset, or not remembering.
                        float defaultSpeed = RememberPlaybackSpeed.defaultSpeed();
                        float speed = Settings.getBoolean("remember_playback_speed", false) && defaultSpeed > 0
                                && VideoInformation.getPlaybackSpeed() != defaultSpeed ? defaultSpeed : 1.0f;
                        VideoInformation.overridePlaybackSpeed(speed);
                        return true;
                    });
            updateText(VideoInformation.getPlaybackSpeed());
            Logger.debug(() -> "Playback speed dialog button initialized");
        } catch (Exception exception) {
            Logger.error(() -> "Playback speed dialog button initialization failure: " + exception);
        }
    }

    /** Injection point, when the playing speed changes. */
    public static void videoSpeedChanged(float speed) {
        updateText(speed);
    }

    private static synchronized void updateText(float speed) {
        if (button != null) button.setTextOverlay(speedFormatter.format(speed));
    }
}

File: apps/youtube/patch/src/main/kotlin/app/reseam/patches/youtube/video/speed/PlaybackSpeedDialogButton.kt
// SPDX-FileCopyrightText: 2026 AunAli K. <hello@auna.li>
// SPDX-FileCopyrightText: 2026 ReVanced contributors
// SPDX-License-Identifier: GPL-3.0-or-later

package app.reseam.patches.youtube.video.speed

import app.reseam.patch.ExtClass
import app.reseam.patch.Type
import app.reseam.patch.patch
import app.reseam.patch.settings.section
import app.reseam.patches.youtube.core.YOUTUBE
import app.reseam.patches.youtube.core.YouTubeSettings
import app.reseam.patches.youtube.core.YouTubeSettingsPages
import app.reseam.patches.youtube.core.youTubeSettings
import app.reseam.patches.youtube.internal.playerControls
import app.reseam.patches.youtube.internal.registerPlayerControlInitializer
import app.reseam.patches.youtube.internal.registerPlayerControlLayout
import app.reseam.patches.youtube.internal.userSelectedPlaybackSpeedHook
import app.reseam.patches.youtube.internal.videoInformationHook
import app.reseam.patches.youtube.internal.videoSpeedChangedHook

object PlaybackSpeedDialogButton : ExtClass("app.reseam.youtube.speed.PlaybackSpeedDialogButton") {
    val initialize = static("initialize", Type.View)
    val videoSpeedChanged = static("videoSpeedChanged", Type.Float)
}

val playbackSpeedDialogButton = patch("Playback speed player button") {
    description("Adds a button that opens the configured playback-speed picker.")
    compatibleWith(YOUTUBE)
    dependsOn(youTubeSettings, playerControls, customPlaybackSpeed, videoInformationHook)
    settings(youTubeSettings, section(YouTubeSettingsPages.Controls, "Player buttons", YouTubeSettings.playbackSpeedDialogButton))

    execute {
        resources.addId("reseam_playback_speed_button")
        resources.addFile(
            "drawable",
            "reseam_playback_speed",
            "res/drawable/reseam_playback_speed.xml",
            PlaybackSpeedDialogButton::class.java.getResourceAsStream("/buttons/reseam_playback_speed.xml")!!.readBytes(),
        )
        registerPlayerControlLayout("/buttons/playback_speed.xml", "reseam_playback_speed_button_container", position = 4)
        registerPlayerControlInitializer(PlaybackSpeedDialogButton.initialize)
        videoSpeedChangedHook(PlaybackSpeedDialogButton.videoSpeedChanged)
        userSelectedPlaybackSpeedHook(PlaybackSpeedDialogButton.videoSpeedChanged)
    }
}
